File: combination-sum/yeonguchoe.java
class Solution {
    // Time complexity: O(candidate^(target/smallest candidate))
    // Space complexity: O(target/smallest candidate)

    List<List<Integer>> result = new ArrayList<>();

    public void backtracking(int remainingAmount, List<Integer> combination, int startPoint, int[] candidateList,
            List<List<Integer>> resultList) {
        if (remainingAmount == 0) {
            resultList.add(new ArrayList<>(combination));
            return;
        }

        if (remainingAmount < 0) {
            return;
        }

        for (int i = startPoint; i < candidateList.length; i++) {
            combination.add(candidateList[i]);
            backtracking(remainingAmount - candidateList[i], combination, i, candidateList, resultList);
            combination.remove(combination.size() - 1);
        }
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        ArrayList<Integer> initialCombination = new ArrayList<>();
        backtracking(target, initialCombination, 0, candidates, result);
        return result;
    }
}

File: unique-paths/yeonguchoe.cpp
class Solution {
public:
    int combination(int n, int r) {
        int k = n - r > r ? r : n - r;
        unsigned long long result = 1;
        for (int i = 0; i < k; i++) {
            result = result * (n - i) / (i + 1);
        }
        return result;
    }
    int uniquePaths(int m, int n) { return combination(m - 1 + n - 1, m - 1); }
    // 시간 복잡도: O(min(m,n))
    // 공간 복잡도: O(1)
};
